### PR TITLE
endpoint: fix policy drops on first regeneration

### LINKED ISSUE #28645: Policy drops for kube-apiserver entity during first regeneration after agent restart
## Summary

When an endpoint references uses a CiliumNetworkPolicy that selects the `kube-apiserver` entity, and that entity is outside of the cluster (e.g. cloud provider), then traffic can be dropped while the endpoint is undergoing regeneration.

This is due to a complicated interaction between CIDR identity restoration and policy updates.

There is a broader attempt to fix this in #3897 but that effort has stalled due to complications interacting with Envoy.

## Background

The prerequisites for this issue are:

1. The apiserver is external to the cluster (e.g. on a cloud provider)
2. A pod selects the apiserver via the CNP entity.
3. An agent is restarted

## Step-by-step

**Step 0**

The pod selects the `reserved:kube-apiserver` label. This resolves to a CIDR identity - the one created for the apiserver. Life is good:
```
root@kind-worker:/home/cilium# cilium identity list
16777217   cidr:172.25.0.2/32
           reserved:kube-apiserver
           reserved:world-ipv4
root@kind-worker:/home/cilium# cilium bpf policy get -n 323
POLICY   DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   AUTH TYPE   BYTES   PACKETS   PREFIX      
Allow    Egress      16777217   ANY          NONE         disabled    10794   86        0    
```

**Step 1**

The agent is restarted. Early on in the restart process, the CIDR identity is restored. Critically, it _does not contain the reserved:kube-apiserver` label:
```
root@kind-worker:/home/cilium# cilium identity list
16777217   cidr:172.25.0.2/32
           reserved:world-ipv4
root@kind-worker:/home/cilium# cilium bpf policy get -n 323
POLICY   DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   AUTH TYPE   BYTES   PACKETS   PREFIX      
Allow    Egress      16777217   ANY          NONE         disabled    10794   86        0   
```

Being unregenerated, the pod still uses the old ipcache, which maps `172.25.0.2 -> 16777217`

**Step 2**

The kubernetes caches are synced. A new identity is created in response to the `default/kubernetes` service:
```
root@kind-worker:/home/cilium# cilium identity list
16777217   cidr:172.25.0.2/32
           reserved:world-ipv4
16777218   cidr:172.25.0.2/32
           reserved:kube-apiserver
           reserved:world-ipv4
root@kind-worker:/home/cilium# cilium bpf policy get -n 323
POLICY   DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   AUTH TYPE   BYTES   PACKETS   PREFIX      
Allow    Egress      16777217   ANY          NONE         disabled    10794   86        0   
```

The new ipcache now maps `172.25.0.2 -> 16777218`, but **the pod does not yet reference it** as it is unregenerated. Thus, the pod's ipcache maps `172.25.0.2 -> 16777217`. So, traffic is allowed.

**Step 3**
Endpoint regeneration starts. Policy is calculated. Crucially, long before BPF is being updated, we call `ep.syncPolicyMap()` which updates identities to those in the *new* ipcache:
```
root@kind-worker:/home/cilium# cilium identity list
16777217   cidr:172.25.0.2/32
           reserved:world-ipv4
16777218   cidr:172.25.0.2/32
           reserved:kube-apiserver
           reserved:world-ipv4
root@kind-worker:/home/cilium# cilium bpf policy get -n 323
POLICY   DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   AUTH TYPE   BYTES   PACKETS   PREFIX      
Allow    Egress      16777218   ANY          NONE         disabled    10794   86        0   
```

So, the pod's policymap only allows 16777218 but its ipcache maps `172.25.0.2 -> 16777217`. So, traffic is **incorrectly dropped!**

**Step 4**
Endpoint regeneration proceeds. It does some bpf compliation. It dreams of electric sheep. Traffic is still being incorrectly dropped. This can take some number of seconds on busy nodes with a lot of pods.

**Step 5**
Eventually, the new bpf program is loaded. It references the new ipcache, which maps `172.25.0.2 -> 16777218` and traffic is allowed.

**Step 6**
For various details surrounding Envoy, the endpoint does a second `ep.syncPolicyMap()`.

## Proposed solution
My first proposed solution would be to skip removing entries from the policy map on the first `ep.syncPolicyMap()` of the first regeneration. That would prevent spurious policy drops.

It is possible this could cause new policy drops in the event of deny policies. For this to happen, an IP address would have to switch from denied to allowed during the agent restart. So, very, very unlikely.

The open question would be: how does this affect L7 policies? Could we somehow have a stale entry that causes issues?

### PR DESCRIPTION
**Pull request by squeed**
There are two small changes in this PR.

The first blocks endpoint regeneration until the ipcache has written at least once. This is an unlikely race, but important to make explicit.

The second commit fixes a source of policy drops on initial regeneration. Because an un-regenerated endpoint references an old ipcache (because the map is referenced by file descriptor), IPs still have "old" identities until a new BPF program is applied.

However, the endpoint's policy map is synchronized *before* the new bpf program is created. That means that if any IPs have changed identities after restart, the policy map will reference the new IP, but the endpoint's ipcache provides the old identity, leading to a policy drop.

The solution is to defer removing stale entries until after the endpoint has been regenerated for the first time. Fortunately the code already has two calls to `ep.syncPolicyMap()`. We just need to wire up the bits to keep stale entries as applicable.

Fixes: #28645
Related: #3897 

```release-note
Fixes a bug where polices that allow access to the kubernetes apiserver may experience brief interruptions on agent restart
```

### CODE REVIEW

File: daemon/cmd/daemon_main.go [hunk 1]
```diff
@@ -1743,6 +1743,13 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
 		<-params.CacheStatus
 	}
 	bootstrapStats.k8sInit.End(true)
+
+	// Wait for the ipcache to write at least once before restoring.
+	// Note that this must happen after waiting for k8s to sync, as the ipcache
+	// also blocks until k8s caches have synced.
+	log.Info("Waiting for ipcache to complete initial write")
+	d.ipcache.WaitForFirstLabelInjection()
```
Comment (mhofstetter) [inline daemon/cmd/daemon_main.go]: AFAIU, this synchronous wait not only blocks the restoration but all of the following tasks in the `startDaemon` function - and therefore might unnecessarily increase the startup time of the agent.

Since the restoration itself is also executed asynchronously in its own go-routine(s), wouldn't it be nicer to use an the asynchronous mechanism (and signal via channel) to wait for the first ipcache write too? This way we would only make the dependent "tasks" waiting for this.

**Review by mhofstetter — COMMENTED** [review]
thanks for this @squeed 

i have a question / input for the agent related bits.

File: daemon/cmd/daemon_main.go (hunk 1 above)
Comment (squeed) [reply daemon/cmd/daemon_main.go]: Sure, makes total sense. Very good point.

**Comment by squeed** [comment]
@jrajahalme. I don't understand enough of the details around l7 proxy ports to know if this will break things. I don't *think* so, but I'm not so sure.

**Review by mhofstetter — APPROVED** [review]
LGTM (sig-agent part).  thanks for addressing my concern.

**Comment by squeed** [comment]
I need to do some testing to see how far back this bug goes; I suspect it has been present for some time.

File: pkg/ipcache/metadata.go [hunk 2]
```diff
@@ -179,6 +179,9 @@ func (m *metadata) getLocked(prefix netip.Prefix) PrefixInfo {
 // Returns the CIDRs that were not yet processed, for example due to an
 // unexpected error while processing the identity updates for those CIDRs
 // The caller should attempt to retry injecting labels for those CIDRs.
+//
+// Note: do not call this function (except in tests)! Rather, calling
+// UpsertLabels() / UpsertMetadata() will trigger an update asynchronously.
 func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.Prefix) (remainingPrefixes []netip.Prefix, err error) {
```
Comment (joestringer) [inline pkg/ipcache/metadata.go:185]: Incidentally we can probably hide this with a rename to lowercase this as well as  `TriggerLabelInjection()` to prevent such misuse.

One of these days we should probably rework the diagram as well, since there are more and more sources of information using this model. But I digress.

These suggestions are unrelated to the PR, can be submitted separately or as a separate commit.

File: pkg/ipcache/metadata_test.go [hunk 3]
```diff
@@ -413,9 +413,9 @@ func TestInjectWithLegacyAPIOverlap(t *testing.T) {
 	identityReferences++
 	for i := 0; i < 2; i++ {
 		IPIdentityCache.UpsertLabels(prefix, labels, source.CustomResource, resource)
-		// Need to wait for the label injector to finish; easiest just to remove it
-		IPIdentityCache.controllers.RemoveControllerAndWait(LabelInjectorName)
 	}
+	// Need to wait for the label injector to finish; easiest just to remove it
+	IPIdentityCache.ShutdownLabelInjection()
```
Comment (joestringer) [inline pkg/ipcache/metadata_test.go:418]: Was this happening inside the loop for a reason? :thinking: 

Given that this particular test is about testing reference counts and this has an arbitrary `2` for inserting the exact same labels, I think this is trying to ensure that the inner function completes twice. However if this line is moved outside the loop then we don't know if the inner `InjectLabels()` runs twice.

Conclusion: This *must* ensure that the inner loop in `InjectLabels()` runs twice. This change as-is breaks the test.

File: pkg/ipcache/ipcache.go [hunk 4]
```diff
@@ -128,6 +129,11 @@ type IPCache struct {
 	// references to identities and removing the corresponding IPCache
 	// entries if unused.
 	deferredPrefixRelease *asyncPrefixReleaser
+
+	// labelsInjected is closed when the first label injection is complete.
+	labelsInjected       chan struct{}
+	labelsInjectedClosed bool      // because closing channels is not idempotent
```
Comment (joestringer) [inline pkg/ipcache/ipcache.go:135]: nit: The bool might be tempting to use but it's not using atomics or some other thread-safe accessor. Should we document this or swap this one also to a `sync.Once`?

```suggestion
	labelsInjectedClosed bool      // because closing channels is not idempotent. Used only by doLabelInjection().
```

Comment (joestringer) [reply pkg/ipcache/ipcache.go:135]: Ah I guess from the test code you'd need to reset the `sync.Once` to make it "more than once" if following that approach :)

File: daemon/cmd/state.go [hunk 5]
```diff
@@ -342,6 +342,11 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
 		}
 		log.WithField(logfields.EndpointID, ep.ID).Info("Successfully restored endpoint. Scheduling regeneration")
 		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
+
+			// Wait for the ipcache to write at least once before regenerating.
+			// Otherwise, there could be transient policy drops.
+			d.ipcache.WaitForFirstLabelInjection()
```
Comment (joestringer) [inline daemon/cmd/state.go:348]: This case makes sense.

Does the base programs generation need to likewise block? Or is there already some ordering constraint there somehow, for instance because the daemon does the `initMaps()` strictly before ipcache starts based on the current ordering, and therefore the new ipcache map is already created when the ipcache starts generating entries? I *think* that's the case, so no need.. but I did not verify.

As far as I understand this will also indirectly block each endpoint regeneration on syncing all k8s state. I think this is probably the correct behaviour but may surprise folks in big clusters.

File: pkg/endpoint/bpf.go [hunk 6]
```diff
@@ -847,7 +847,14 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 		// GH-3897 would fix this by creating a new map to do an atomic swap
 		// with the old one.
 		stats.mapSync.Start()
-		err = e.syncPolicyMap()
+		// If we're doing out first regeneration after a restore, the endpoint is still
+		// referencing the old ipcache at this point in time. Thus, we should only add
+		// new entries in to the policy map without removing stale entries. We will
+		// clean up the stale entries at the end of regeneration.
+		// Once the endpoint is referencing the new ipcache, this is no longer necessary.
+		// See: GH-28645
```
Comment (joestringer) [inline pkg/endpoint/bpf.go:855]: Hmm, I can't help but wonder if the better option is to just delete the old policymap and recreate it ~65 lines up at `policymap.OpenOrCreate()`. The existing datapath will continue to operate, and hold the reference to the existing policymap, using its content, meanwhile we unpin that version and create / pin a new policymap which this code can then happily write whatever it wants without any repercussions. The new one would only be plumbed as part of the upcoming BPF setup steps. No new complexity from understanding whether this code has partially calculated or fully synced the desired state into the maps, and no delays to correct enforcement.

Comment (joestringer) [reply pkg/endpoint/bpf.go:855]: If we think it's better to continue with the current approach, I would like to see a more comprehensive description of why adding new entries into the map is OK and deleting entries is not OK. I think in principle the idea is that new policymap entries are allowing traffic which should now be allowed, and if we don't delete the old entries for previously-used identities then those will continue to flow. But equally it's possible for the policy calculation to inject new deny rules that now match newly allocated identities and that could cause a subtly different form of the policy drops. More of a corner case, harder to track down.

Furthermore if this doesn't delete the old entries then yes per the commit message it needs to follow up and do that the second time... are we just relying on the periodic policymap sync controller to do that? So if it fails to run or gets delayed or future developers change it, then they could overlook that this extends the period of policy not being quite 100% in sync with the intended policy (eventually consistent, sure, but desynced longer than necessary).

**Review by joestringer — CHANGES_REQUESTED** [review]
I found some minor nits in the first commit, but the most notable questions come in at the end of my review on particularly around why we reuse the same policymap.

Is there a way to test this? I don't see any test code changes.

The ginkgo failure seems a bit suspect. If endpoint list CLI cannot return, that can mean that the agent has deadlocked or is otherwise somehow depriving the CLI from being able to successfully make API calls:

```
cmd: kubectl exec -n kube-system cilium-nngml -c cilium-agent -- cilium-dbg endpoint list
Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 Error: cannot get endpoint list: Cilium API client timeout exceeded
	 
	 command terminated with exit code 1
```

File: pkg/endpoint/bpf.go (hunk 6 above)
Comment (jrajahalme) [reply pkg/endpoint/bpf.go:855]: IMO creating a new policy map instead of updating the old one as Joe suggests here would be the right thing to do... im my mental model we already did that :-)

**Comment by squeed** [comment]
Switching to draft while I think about the feedback; if swapping to a new policy map on restart is possible, that seems like the better idea.
